### ISSUE TITLE
add possibility to initialize GLM with "design matrices" (or events)

### ISSUE THREAD

**Issue by Remi-Gau** [nilearn/nilearn#4505]
The rationale would be to be able to generate a report with the model to be fitted.

So something like:

```python
first_level_model = FirstLevelModel(t_r, events=events)
# inspect that everything is OK
report = first_level_model.generate_report()
# then fit the model
first_level_model = first_level_model.fit(fmri_img)
```

Instead of the current:

```python
first_level_model = FirstLevelModel(t_r, events=events)
# fit the model
first_level_model = first_level_model.fit(fmri_img, events=events)
# inspect that everything was OK but if anything was wrong we need to refit everything
report = first_level_model.generate_report()
```

Not sure if this is easily feasible but in terms of using the reports to do early QC it would be better from a user experience perspective.

**Comment by Remi-Gau** [comment]
from a quick look it seems that it would be possible to do it

@bthirion do you think it would be something desirable to have / do ?

**Comment by bthirion** [comment]
Normally, you want to have events associated with fMRI data. Both are provided at fit time.
You want to specify the GLM irrespective of the events.
Home I got your point right.
Best,

**Comment by Remi-Gau** [comment]
> Normally, you want to have events associated with fMRI data. Both are provided at fit time.
> You want to specify the GLM irrespective of the events.

I want to be able to specify the model irrespective of the data.

I think that it'd be nice to be able to distinguish model specification (which is fast and allows to review / inspect the model) from model estimation.

From my understanding SPM and FSL, do make that distinction by allowing inspection of the specified model before fitting. AFNI I think does not allow that.

From what I can read FSL even allows to pass contrasts with model specification which allows to compute design efficiency as well (since this does not require any data).

Not exactly a hill I want to die on but this is not urgent either so I will leave the issue open in case some other people are interested.

Could also be discussed at a dev meeting.

**Comment by bthirion** [comment]
+1 for discussion. 
Nilearn offers the possibility to build design matrices, but indeed, the API is different which makes it less easy to run. 
On the other hand, I want to insist that GLM specification involves a fit(), so that it would make little sense to me to provide events descriptors during model setup. Nothing is carved in stone though, so I'm happy to hear more about your motivation.

**Comment by Remi-Gau** [comment]
Now that I have a better understanding about sklearn estimators "good practices", I would suggest to maybe do something like this instead

```python
first_level_model = FirstLevelModel(t_r)
first_level_model = first_level_model.fit(run_imgs=None, events=events, design_only=True)

# inspect model that everything is OK
report = first_level_model.generate_report()

# actually run model
first_level_model.fit(run_imgs=run_imgs, events=events)
```

**Comment by bthirion** [comment]
Why not provide `design_only` at instantiation time rather than fit time ?

**Comment by Remi-Gau** [comment]
oh yeah that'd fine too but it would add another line to 'unset' it before doing a fit for real:

```python
first_level_model = FirstLevelModel(t_r, design_only=True)
first_level_model = first_level_model.fit(run_imgs=None, events=events)

# inspect model that everything is OK
report = first_level_model.generate_report()

# unset
first_level_model.design_only = False

# actually run model
first_level_model.fit(run_imgs=run_imgs, events=events)
```

**Comment by bthirion** [comment]
Well, the idea is that unsetting/resetting should be rare. To me you rather what to have another GLM if you want to input events and params instead of design matrices, or vice versa ?

**Comment by Remi-Gau** [comment]
yeah that's another possibility though I don't think we want to go as far as making it impossible to change estimators attributes

**Comment by bthirion** [comment]
Agreed.
